cranelift-codegen/src/simple_preopt.rs

```diff
@@ -516,6 +517,27 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
                     imm,
                 } = pos.func.dfg[iconst_inst]
                 {
+                    // For signed comparisons, we may need to cast the immediate to the instruction
+                    // result (controlling) type.
+                    let imm = match cond {
+                        IntCC::SignedLessThan
+                        | IntCC::SignedLessThanOrEqual
+                        | IntCC::SignedGreaterThan
+                        | IntCC::SignedGreaterThanOrEqual => {
```

Comment:
It's not immediately clear why we need to do this explicit sign-extend step. The `iconst` should already have the same type as the comparison's controlling type. Is it so that `Imm64` values outside the range of that type get normalized? If so, it's unclear why we need to do that for signed and not unsigned here.

Reply:
See https://github.com/CraneStation/cranelift/issues/815

Reply:
Good question: the reason is to do so only for the sign of the immediate, since it's always stored as an i64. When we're generating the instruction, on x86 we use encodings `icscc_ib` or `icscc_id`, which do an actual cast of the immediate from i64 to u8 or u32. So for unsigned immediates this shouldn't be needed.